File: zio-8807/src/mpsc/vyukov/Padded128.java
package mpsc.vyukov;

import java.io.Serializable;
import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;

import mpsc.Queue;

abstract class Padded128ClassPad implements Serializable {
	protected int _0;
	protected long _1;
	protected long _2;
	protected long _3;
	protected long _4;
	protected long _5;
	protected long _6;
	protected long _7;
	protected long _8;
	protected long _9;
	protected long _a;
	protected long _b;
	protected long _c;
	protected long _d;
	protected long _e;
}

abstract class Padded128Write extends Padded128ClassPad {
	protected transient volatile Padded128.Node write;
}

abstract class Padded128WritePad extends Padded128Write {
	protected int __0;
	protected long __1;
	protected long __2;
	protected long __3;
	protected long __4;
	protected long __5;
	protected long __6;
	protected long __7;
	protected long __8;
	protected long __9;
	protected long __a;
	protected long __b;
	protected long __c;
	protected long __d;
	protected long __e;
	protected long __f;
}

public final class Padded128<A> extends Padded128WritePad implements Queue<A> {

	private transient Node read;

	public Padded128() {
		read = write = new Node(null);
	}

	@Override
	public void add(A data) {
		Node next = new Node(data);
		Node prev = WRITE.getAndSet(this, next);
		Node.NEXT.lazySet(prev, next);
	}

	@SuppressWarnings("unchecked")
	@Override
	public A poll() {
		Node next = read.next;

		if (next == null)
			return null;

		read = next;
		Object data = next.data;
		next.data = null;
		return (A) (data);
	}

	static class Node {
		Object data;
		volatile Node next;

		Node(Object data) {
			this.data = data;
		}

		Node(Object data, Node next) {
			this.data = data;
			this.next = next;
		}

		static final AtomicReferenceFieldUpdater<Node, Node> NEXT = AtomicReferenceFieldUpdater.newUpdater(Node.class,
				Node.class, "next");
	}

	static final AtomicReferenceFieldUpdater<Padded128Write, Node> WRITE = AtomicReferenceFieldUpdater
			.newUpdater(Padded128Write.class, Node.class, "write");
}

File: zio-8807/src/mpsc/custom/Padded128.scala
package mpsc
package custom

import java.util.concurrent.atomic.AtomicReference

private[custom] abstract class Padded128Write[A]
    extends AtomicReference(List.empty[A])
    with Serializable

private[custom] abstract class Padded128WritePad[A] extends Padded128Write[A] {
  protected val _0 = 0L;
  protected val _1 = 0L;
  protected val _2 = 0L;
  protected val _3 = 0L;
  protected val _4 = 0L;
  protected val _5 = 0L;
  protected val _6 = 0L;
  protected val _7 = 0L;
  protected val _8 = 0L;
  protected val _9 = 0L;
  protected val _a = 0L;
  protected val _b = 0L;
  protected val _c = 0L;
  protected val _d = 0L;
}

final class Padded128[A] extends Padded128WritePad[A] with Queue[A] {

  @transient private var read = List.empty[A]

  def add(head: A): Unit = {
    var tail = null.asInstanceOf[List[A]]
    var prev = get()
    do {
      tail = prev
      prev = compareAndExchange(tail, head :: tail)
    } while (prev ne tail)
  }

  def poll(): A = {
    var read = this.read
    if (read.isEmpty) {
      read = getAndSet(Nil)
      if (read eq Nil) return null.asInstanceOf[A]
      if (read.tail ne Nil) read = read.reverse
    }
    this.read = read.tail
    read.head
  }
}
